### PR TITLE
CR-1127142 shim: add flag to exit dev_offline status check loop forcibly

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

#### Problem solved by the commit
xbutil reset test hangs indefinitely on Microsoft Azure multi-cards test environment when multiple tests run simultaneously.
It is due to device is not coming online after reset.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
CR-1127142 This PR fixes indefinite run in while loop issue in reset sequence.
#### How problem was solved, alternative solutions (if any) and why they were rejected
Added a timer in shim driver, which exits the while loop if the timer reaches 120 seconds (default). One can set to any value using xrt runtime config file, for example, set value to 100 seconds
Runtime.dev_offline_timer=100
#### Risks (if any) associated the changes in the commit
NA
#### What has been tested and how, request additional testing if necessary
xbutil reset, xbutil validate
#### Documentation impact (if any)
NA

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
I will add Stephen and Soren once the change approved by Vamshi & Arpit.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 1]
```diff
@@ -1110,10 +1110,19 @@ int shim::resetDevice(xclResetKind kind)
 
     dev_fini();
 
+    unsigned int timer = xrt_core::config::get_device_offline_timer();
+    std::chrono::time_point<std::chrono::system_clock> start, end;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1114]:
```suggestion
```

File: src/runtime_src/core/common/config_reader.h [hunk 2]
```diff
@@ -831,6 +831,15 @@ get_flag_sw_emu_kernel_debug()
   return value;
 }
 
+// This flag is added to exit device offline status check loop forcibly.
+// By default, device offline status loop runs for 60 seconds.
+inline bool
```
Comment (stsoe) [inline src/runtime_src/core/common/config_reader.h:836]:
```suggestion
inline unsigned int
```

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 3]
```diff
@@ -1110,10 +1110,19 @@ int shim::resetDevice(xclResetKind kind)
...
+    start = std::chrono::system_clock::now();
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1115]:
```suggestion
    auto start = std::chrono::system_clock::now();
```

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 4]
```diff
@@ -1110,10 +1110,19 @@ int shim::resetDevice(xclResetKind kind)
...
     while (dev_offline) {
         std::this_thread::sleep_for(std::chrono::milliseconds(500));
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1117]: Magic constant and no comments.  This logic, regardless of magic or not, must be commented; it is not obvious.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 5]
```diff
@@ -1110,10 +1110,19 @@ int shim::resetDevice(xclResetKind kind)
...
         pcidev::get_dev(mBoardNumber)->sysfs_get<int>("",
             "dev_offline", err, dev_offline, -1);
+        end = std::chrono::system_clock::now();
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1120]:
```suggestion
        auto end = std::chrono::system_clock::now();
```

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 6]
```diff
@@ -1110,10 +1110,19 @@ int shim::resetDevice(xclResetKind kind)
...
+        std::chrono::duration<double> elapsed_seconds = end - start;
+        if (elapsed_seconds.count() > timer) {
+            xrt_logmsg(XRT_WARNING, "%s: device unable to come online, so exiting", __func__);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1123]: exiting what?   breaking the while loop, but otherwise just proceeding as normal.

**Review by stsoe — CHANGES_REQUESTED** [review]
Please update the PR description and fill out all the items.

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by arpitxilinx — APPROVED** [review]
@rajkumar-xilinx I am assuming you or Durga have tested this changes and now xbutil at least exits. I think the underlying issue of why reset is failing still needs to be debugged.

**Comment by rajkumar-xilinx** [comment]
> @rajkumar-xilinx I am assuming you or Durga have tested this changes and now xbutil at least exits. I think the underlying issue of why reset is failing still needs to be debugged.

Hi @arpitxilinx yes, we validated this change.

**Comment by rajkumar-xilinx** [comment]
retest this please

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by rajkumar-xilinx** [comment]
Hi @stsoe , could you please review this change ? thank you.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 7]
```diff
@@ -1098,27 +1098,44 @@ int shim::xclGetDeviceInfo2(xclDeviceInfo2 *info)
...
 
     dev_fini();
 
+    /*
+     * This loop used to wait for device come online and ready to use. It uses "dev_offline"
+     * sysfs node to retrieve latest status of device in the interval of 500 milli sec.
+     * In certain testing environement, reset never complete and waits indefinitely
+     * in this loop for dev_offline status to be false. To overcome this infinite loop issue,
+     * added loop_timer (default value set to 120 sec) which is used to exit the loop.
+     */
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1121]: C++ inline commenting style is `//`.  Please follow the convention as much as possible.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 8]
```diff
@@ -1098,27 +1098,44 @@ int shim::xclGetDeviceInfo2(xclDeviceInfo2 *info)
...
+    int ret = 0;
+
     // Only XCL_USER_RESET is supported on user pf.
     if (kind != XCL_USER_RESET)
         return -EINVAL;
 
     std::string err;
     int dev_offline = 1;
-    int ret = mDev->ioctl(mUserHandle, DRM_IOCTL_XOCL_HOT_RESET);
+    ret = mDev->ioctl(mUserHandle, DRM_IOCTL_XOCL_HOT_RESET);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:1109]: Why was declaration moved to 1101?  Declare and defined variables where they are first used.

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by stsoe — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by chvamshi-xilinx** [comment]
@rajkumar-xilinx , Cherry pick this to 2022.1

**Pull request merged**